File: sdk-py/smolvm/smolvm_client.py
"""smolvm Python SDK — SmolvmClient.

Top-level client. The main entry point for the SDK.
Creates sandboxes, microvms, and fleets.
"""

from __future__ import annotations

from typing import Any

from .client import SmolvmHttpClient
from .sandbox import Sandbox
from .microvm import MicroVM
from .fleet import SandboxFleet, create_fleet
from .types import (
    CheckpointMetadata,
    CreateMicroVMOptions,
    CreateSandboxOptions,
    HealthResponse,
    MicroVMInfo,
    SandboxInfo,
    SnapshotInfo,
    StarterInfo,
)


class SmolvmClient:
    """Top-level smolvm client.

    Args:
        base_url: smolvm server URL. Defaults to SMOLVM_URL env var
                  or http://127.0.0.1:8080.
        api_token: Bearer token for authentication. Defaults to
                   SMOLVM_API_TOKEN env var.
    """

    def __init__(self, base_url: str | None = None, api_token: str | None = None):
        self._http = SmolvmHttpClient(base_url, api_token=api_token)

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._http.close()

    # --------------------------------------------------------------------------
    # Health
    # --------------------------------------------------------------------------

    async def health(self) -> HealthResponse:
        return await self._http.health()

    # --------------------------------------------------------------------------
    # Sandboxes
    # --------------------------------------------------------------------------

    async def create(
        self,
        name: str,
        cpus: int = 2,
        memory_mb: int = 1024,
        network: bool = False,
        overlay_gb: int | None = None,
        storage_gb: int | None = None,
        mounts: list[Any] | None = None,
        ports: list[Any] | None = None,
        init_commands: list[str] | None = None,
        allowed_domains: list[str] | None = None,
        default_user: str | None = None,
        from_starter: str | None = None,
        secrets: list[str] | None = None,
    ) -> Sandbox:
        """Create a new sandbox (ephemeral VM). Does NOT start it.

        Args:
            secrets: Secret names to inject via the secret proxy
                     (e.g., ["anthropic", "openai"]). Requires secrets
                     configured on the server with ``--secret name=value``.
                     The sandbox gets ``*_BASE_URL`` env vars pointing to a
                     local proxy and placeholder API keys. Real keys never
                     enter the VM.
        """
        req: dict[str, Any] = {
            "name": name,
            "resources": {
                "cpus": cpus,
                "memory_mb": memory_mb,
                "network": network,
            },
        }
        if overlay_gb is not None:
            req["resources"]["overlay_gb"] = overlay_gb
        if storage_gb is not None:
            req["resources"]["storage_gb"] = storage_gb
        if allowed_domains is not None:
            req["resources"]["allowed_domains"] = allowed_domains
        if mounts is not None:
            req["mounts"] = mounts
        if ports is not None:
            req["ports"] = ports
        if init_commands is not None:
            req["init_commands"] = init_commands
        if default_user is not None:
            req["default_user"] = default_user
        if from_starter is not None:
            req["from_starter"] = from_starter
        if secrets is not None:
            req["secrets"] = secrets

        await self._http.create_sandbox(req)
        return Sandbox(name, self._http)

    async def create_and_start(
        self,
        name: str,
        cpus: int = 2,
        memory_mb: int = 1024,
        network: bool = False,
        overlay_gb: int | None = None,
        storage_gb: int | None = None,
        mounts: list[Any] | None = None,
        ports: list[Any] | None = None,
        init_commands: list[str] | None = None,
        allowed_domains: list[str] | None = None,
        default_user: str | None = None,
        from_starter: str | None = None,
        secrets: list[str] | None = None,
    ) -> Sandbox:
        """Create and immediately start a sandbox."""
        sandbox = await self.create(
            name,
            cpus=cpus,
            memory_mb=memory_mb,
            network=network,
            overlay_gb=overlay_gb,
            storage_gb=storage_gb,
            mounts=mounts,
            ports=ports,
            init_commands=init_commands,
            allowed_domains=allowed_domains,
            default_user=default_user,
            from_starter=from_starter,
            secrets=secrets,
        )
        await sandbox.start()
        return sandbox

    async def get(self, name: str) -> Sandbox:
        """Get an existing sandbox by name."""
        await self._http.get_sandbox(name)
        return Sandbox(name, self._http)

    async def list(self) -> list[SandboxInfo]:
        return await self._http.list_sandboxes()

    # --------------------------------------------------------------------------
    # Checkpoints
    # --------------------------------------------------------------------------

    async def list_checkpoints(self) -> list[CheckpointMetadata]:
        return await self._http.list_checkpoints()

    async def restore_checkpoint(self, checkpoint_id: str, name: str) -> Sandbox:
        """Restore a checkpoint into a new sandbox."""
        await self._http.restore_checkpoint(checkpoint_id, name)
        return Sandbox(name, self._http)

    async def delete_checkpoint(self, checkpoint_id: str) -> None:
        await self._http.delete_checkpoint(checkpoint_id)

    # --------------------------------------------------------------------------
    # Starters
    # --------------------------------------------------------------------------

    async def list_starters(self) -> list[StarterInfo]:
        """List available starter templates."""
        return await self._http.list_starters()

    # --------------------------------------------------------------------------
    # Snapshots
    # --------------------------------------------------------------------------

    async def list_snapshots(self) -> list[SnapshotInfo]:
        """List available snapshots."""
        return await self._http.list_snapshots()

    async def pull_snapshot(self, snapshot_name: str, sandbox_name: str) -> Sandbox:
        """Pull a snapshot into a new sandbox."""
        await self._http.pull_snapshot(snapshot_name, sandbox_name)
        return Sandbox(sandbox_name, self._http)

    async def delete_snapshot(self, name: str) -> None:
        """Delete a snapshot."""
        await self._http.delete_snapshot(name)

    # --------------------------------------------------------------------------
    # MicroVMs
    # --------------------------------------------------------------------------

    async def create_microvm(
        self,
        name: str,
        cpus: int = 2,
        memory_mb: int = 1024,
        network: bool = False,
        overlay_gb: int | None = None,
        storage_gb: int | None = None,
        mounts: list[Any] | None = None,
        ports: list[Any] | None = None,
    ) -> MicroVM:
        """Create a new MicroVM (persistent VM)."""
        req: dict[str, Any] = {
            "name": name,
            "cpus": cpus,
            "memoryMb": memory_mb,
            "network": network,
        }
        if overlay_gb is not None:
            req["overlay_gb"] = overlay_gb
        if storage_gb is not None:
            req["storage_gb"] = storage_gb
        if mounts is not None:
            req["mounts"] = mounts
        if ports is not None:
            req["ports"] = ports

        await self._http.create_microvm(req)
        return MicroVM(name, self._http)

    async def get_microvm(self, name: str) -> MicroVM:
        await self._http.get_microvm(name)
        return MicroVM(name, self._http)

    async def list_microvms(self) -> list[MicroVMInfo]:
        return await self._http.list_microvms()

    # --------------------------------------------------------------------------
    # Fleet
    # --------------------------------------------------------------------------

    async def create_fleet(
        self,
        prefix: str,
        count: int,
        cpus: int = 1,
        memory_mb: int = 512,
        network: bool = True,
        overlay_gb: int | None = None,
        storage_gb: int | None = None,
    ) -> SandboxFleet:
        """Create a fleet of sandboxes. All are created and started."""
        opts = CreateSandboxOptions(
            cpus=cpus,
            memory_mb=memory_mb,
            network=network,
            overlay_gb=overlay_gb,
            storage_gb=storage_gb,
        )
        return await create_fleet(self._http, prefix, count, opts)

File: sdk-ts/smolvm-client.ts
/**
 * smolvm TypeScript SDK — SmolvmClient
 *
 * Top-level client. The main entry point for the SDK.
 * Creates machines, microvms, and fleets.
 */

import { SmolvmHttpClient } from "./client.ts";
import { Machine } from "./machine.ts";
import { MicroVM } from "./microvm.ts";
import { MachineFleet, createFleet } from "./fleet.ts";
import type {
  CheckpointMetadata,
  CreateMicroVMOptions,
  CreateMachineOptions,
  HealthResponse,
  RestoreCheckpointResponse,
  MachineInfo,
  MicroVMInfo,
  SnapshotInfo,
  StarterInfo,
} from "./types.ts";

export class SmolvmClient {
  private readonly http: SmolvmHttpClient;

  /**
   * Create a new smolvm client.
   *
   * @param baseUrl - smolvm server URL. Defaults to SMOLVM_URL env var or http://127.0.0.1:8080
   */
  constructor(baseUrl?: string) {
    this.http = new SmolvmHttpClient(baseUrl);
  }

  // --------------------------------------------------------------------------
  // Health
  // --------------------------------------------------------------------------

  async health(): Promise<HealthResponse> {
    return this.http.health();
  }

  // --------------------------------------------------------------------------
  // Machinees
  // --------------------------------------------------------------------------

  /**
   * Create a new machine (ephemeral VM).
   * Does NOT start it — call machine.start() after.
   *
   * @example
   * const machine = await client.create("my-vm", { cpus: 2, network: true });
   * await machine.start();
   * await machine.sh("echo hello");
   * await machine.cleanup();
   */
  async create(name: string, opts?: CreateMachineOptions): Promise<Machine> {
    await this.http.createMachine({
      name,
      mounts: opts?.mounts,
      ports: opts?.ports,
      resources: {
        cpus: opts?.cpus ?? 2,
        memory_mb: opts?.memoryMb ?? 1024,
        network: opts?.network ?? false,
        overlay_gb: opts?.overlay_gb,
        storage_gb: opts?.storage_gb,
        allowed_domains: opts?.allowed_domains,
        allowed_cidrs: opts?.allowed_cidrs,
      },
      init_commands: opts?.init_commands,
      default_user: opts?.default_user,
      from_starter: opts?.fromStarter,
      secrets: opts?.secrets,
    });
    return new Machine(name, this.http);
  }

  /**
   * Create and immediately start a machine. Convenience method.
   *
   * @example
   * const machine = await client.createAndStart("my-vm", { network: true });
   * await machine.sh("echo hello");
   * await machine.cleanup();
   */
  async createAndStart(name: string, opts?: CreateMachineOptions): Promise<Machine> {
    const machine = await this.create(name, opts);
    await machine.start();
    return machine;
  }

  /**
   * Get an existing machine by name.
   * Does NOT create it — use create() for that.
   */
  async get(name: string): Promise<Machine> {
    // Verify it exists
    await this.http.getMachine(name);
    return new Machine(name, this.http);
  }

  async list(): Promise<MachineInfo[]> {
    return this.http.listMachines();
  }

  // --------------------------------------------------------------------------
  // MicroVMs
  // --------------------------------------------------------------------------

  /**
   * Create a new MicroVM (persistent VM).
   * MicroVMs use a different REST schema than machines.
   */
  async createMicroVM(name: string, opts?: CreateMicroVMOptions): Promise<MicroVM> {
    await this.http.createMicroVM({
      name,
      cpus: opts?.cpus ?? 2,
      memoryMb: opts?.memoryMb ?? 1024,
      network: opts?.network ?? false,
      overlay_gb: opts?.overlay_gb,
      storage_gb: opts?.storage_gb,
      mounts: opts?.mounts,
      ports: opts?.ports,
    });
    return new MicroVM(name, this.http);
  }

  async getMicroVM(name: string): Promise<MicroVM> {
    await this.http.getMicroVM(name);
    return new MicroVM(name, this.http);
  }

  async listMicroVMs(): Promise<MicroVMInfo[]> {
    return this.http.listMicroVMs();
  }

  // --------------------------------------------------------------------------
  // Checkpoints
  // --------------------------------------------------------------------------

  async listCheckpoints(): Promise<CheckpointMetadata[]> {
    return this.http.listCheckpoints();
  }

  async restoreCheckpoint(checkpointId: string, name: string): Promise<Machine> {
    await this.http.restoreCheckpoint(checkpointId, name);
    return new Machine(name, this.http);
  }

  async deleteCheckpoint(checkpointId: string): Promise<void> {
    return this.http.deleteCheckpoint(checkpointId);
  }

  // --------------------------------------------------------------------------
  // Starters
  // --------------------------------------------------------------------------

  async listStarters(): Promise<StarterInfo[]> {
    return this.http.listStarters();
  }

  // --------------------------------------------------------------------------
  // Snapshots
  // --------------------------------------------------------------------------

  async listSnapshots(): Promise<SnapshotInfo[]> {
    return this.http.listSnapshots();
  }

  async pullSnapshot(snapshotName: string, machineName: string): Promise<Machine> {
    await this.http.pullSnapshot(snapshotName, machineName);
    return new Machine(machineName, this.http);
  }

  async deleteSnapshot(name: string): Promise<void> {
    return this.http.deleteSnapshot(name);
  }

  // --------------------------------------------------------------------------
  // Fleet
  // --------------------------------------------------------------------------

  /**
   * Create a fleet of machines. All are created and started.
   * Names: `{prefix}-0`, `{prefix}-1`, ...
   *
   * @example
   * const fleet = await client.createFleet("worker", 3, { network: true });
   * const results = await fleet.execAll("echo hello");
   * await fleet.cleanup();
   */
  async createFleet(
    prefix: string,
    count: number,
    opts?: CreateMachineOptions,
  ): Promise<MachineFleet> {
    return createFleet(this.http, prefix, count, opts);
  }
}
